File: dark_multi/tmux.py
"""tmux session management for dark-multi."""

import os
import shutil
import sys
from pathlib import Path

from .config import DARK_ROOT, TMUX_SESSION, run, warn, error


class Tmux:
    """tmux session management.

    Layout: One window per branch, each with 2 panes (CLI left, claude right)

    Keys:
        Ctrl-b n/p: switch between branch windows
        Ctrl-b o: switch between CLI and claude panes
        Ctrl-b z: zoom current pane (fullscreen toggle)
    """

    @staticmethod
    def is_available() -> bool:
        return shutil.which("tmux") is not None

    @staticmethod
    def session_exists() -> bool:
        if not Tmux.is_available():
            return False
        result = run(["tmux", "has-session", "-t", TMUX_SESSION])
        return result.returncode == 0

    @staticmethod
    def window_exists(name: str) -> bool:
        if not Tmux.session_exists():
            return False
        result = run(["tmux", "list-windows", "-t", TMUX_SESSION, "-F", "#{window_name}"])
        return result.returncode == 0 and name in result.stdout.split("\n")

    @staticmethod
    def create_window(name: str, container_id: str, branch_path: Path = None) -> None:
        """Create a tmux window with CLI + claude panes."""
        if not Tmux.is_available():
            warn("tmux not available, skipping window creation")
            return

        if not Tmux.session_exists():
            # Create session with first window
            run(["tmux", "new-session", "-d", "-s", TMUX_SESSION, "-n", name])
        else:
            # Kill existing window if present, then create new
            if Tmux.window_exists(name):
                run(["tmux", "kill-window", "-t", f"{TMUX_SESSION}:{name}"])
            run(["tmux", "new-window", "-a", "-t", TMUX_SESSION, "-n", name])

        # Left pane: CLI inside container
        run(["tmux", "send-keys", "-t", f"{TMUX_SESSION}:{name}",
             f"docker exec -it -w /home/dark/app {container_id} bash", "Enter"])

        # Split and create right pane: claude on host
        run(["tmux", "split-window", "-h", "-t", f"{TMUX_SESSION}:{name}"])
        workspace = branch_path or DARK_ROOT / name
        run(["tmux", "send-keys", "-t", f"{TMUX_SESSION}:{name}.1",
             f"cd {workspace} && claude", "Enter"])

        # Select left pane (CLI)
        run(["tmux", "select-pane", "-t", f"{TMUX_SESSION}:{name}.0"])

    @staticmethod
    def kill_window(name: str) -> None:
        if Tmux.window_exists(name):
            run(["tmux", "kill-window", "-t", f"{TMUX_SESSION}:{name}"])

    @staticmethod
    def ensure_meta_window() -> None:
        """Create the dark-meta control plane window if it doesn't exist."""
        if Tmux.window_exists("dark-meta"):
            return

        if not Tmux.session_exists():
            # Create session with meta window
            run(["tmux", "new-session", "-d", "-s", TMUX_SESSION, "-n", "dark-meta"])
            # Enable mouse support for scrolling
            run(["tmux", "set-option", "-t", TMUX_SESSION, "-g", "mouse", "on"])
        else:
            run(["tmux", "new-window", "-t", TMUX_SESSION, "-n", "dark-meta"])

        # Move meta window to be first (index 0)
        run(["tmux", "move-window", "-t", f"{TMUX_SESSION}:dark-meta", "-t", f"{TMUX_SESSION}:0"])

        # Left pane (70%): claude in dark-multi directory
        run(["tmux", "send-keys", "-t", f"{TMUX_SESSION}:dark-meta",
             f"cd {DARK_ROOT.parent / 'dark-multi'} && claude", "Enter"])

        # Right pane (30%): quick reference
        run(["tmux", "split-window", "-h", "-p", "30", "-t", f"{TMUX_SESSION}:dark-meta"])

        # Create quick reference content
        ref_text = r'''echo -e "
\033[1m=== DARK MULTI ===\033[0m

\033[1mBranch commands:\033[0m
  multi ls          - list branches
  multi new <name>  - create branch
  multi stop <name> - stop branch
  multi rm <name>   - remove branch
  multi code <name> - open VS Code

\033[1mtmux:\033[0m
  Ctrl-b n/p  - next/prev window
  Ctrl-b w    - list windows
  Ctrl-b o    - switch pane
  Ctrl-b z    - zoom pane
  Ctrl-b d    - detach

\033[1mWindows:\033[0m
  dark-meta   - this control plane
  <branch>    - CLI | claude
"
'''
        run(["tmux", "send-keys", "-t", f"{TMUX_SESSION}:dark-meta.1", ref_text, "Enter"])

        # Select the claude pane
        run(["tmux", "select-pane", "-t", f"{TMUX_SESSION}:dark-meta.0"])

    @staticmethod
    def attach() -> None:
        if not Tmux.is_available():
            error("tmux not installed")
            sys.exit(1)
        if not Tmux.session_exists():
            error("No tmux session. Start a branch first: multi start <branch>")
            sys.exit(1)
        os.execvp("tmux", ["tmux", "attach", "-t", TMUX_SESSION])

File: internal/tmux/tmux.go
// Package tmux provides tmux session management for dark-multi.
package tmux

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/stachu/dark-multi/internal/config"
)

// IsAvailable returns true if tmux is installed.
func IsAvailable() bool {
	_, err := exec.LookPath("tmux")
	return err == nil
}

// BranchSessionName returns the tmux session name for a branch.
func BranchSessionName(branchName string) string {
	return fmt.Sprintf("dark-%s", branchName)
}

// BranchSessionExists returns true if the branch's tmux session exists.
func BranchSessionExists(branchName string) bool {
	if !IsAvailable() {
		return false
	}
	session := BranchSessionName(branchName)
	cmd := exec.Command("tmux", "has-session", "-t", session)
	return cmd.Run() == nil
}

// CreateBranchSession creates a tmux session for a branch with CLI + claude panes.
func CreateBranchSession(branchName string, containerID string, branchPath string) error {
	if !IsAvailable() {
		return fmt.Errorf("tmux not available")
	}

	session := BranchSessionName(branchName)

	// Kill existing session if present
	if BranchSessionExists(branchName) {
		exec.Command("tmux", "kill-session", "-t", session).Run()
	}

	// Create new session
	cmd := exec.Command("tmux", "new-session", "-d", "-s", session)
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to create session: %w", err)
	}

	// Enable mouse support
	exec.Command("tmux", "set-option", "-t", session, "-g", "mouse", "on").Run()

	// Left pane: CLI inside container
	exec.Command("tmux", "send-keys", "-t", session,
		fmt.Sprintf("docker exec -it -w /home/dark/app %s bash", containerID), "Enter").Run()

	// Split and create right pane: claude on host
	exec.Command("tmux", "split-window", "-h", "-t", session).Run()

	workspace := branchPath
	if workspace == "" {
		workspace = filepath.Join(config.DarkRoot, branchName)
	}
	exec.Command("tmux", "send-keys", "-t", fmt.Sprintf("%s.1", session),
		fmt.Sprintf("cd %s && claude", workspace), "Enter").Run()

	// Select left pane (CLI)
	exec.Command("tmux", "select-pane", "-t", fmt.Sprintf("%s.0", session)).Run()

	return nil
}

// KillBranchSession kills a branch's tmux session.
func KillBranchSession(branchName string) error {
	if BranchSessionExists(branchName) {
		session := BranchSessionName(branchName)
		return exec.Command("tmux", "kill-session", "-t", session).Run()
	}
	return nil
}

// BranchHasAttachedClients returns true if branch's session has clients attached.
func BranchHasAttachedClients(branchName string) bool {
	if !BranchSessionExists(branchName) {
		return false
	}
	session := BranchSessionName(branchName)
	out, err := exec.Command("tmux", "list-clients", "-t", session).Output()
	if err != nil {
		return false
	}
	return len(strings.TrimSpace(string(out))) > 0
}

// OpenBranchInTerminal opens a branch's tmux session in a terminal window.
// If already attached somewhere, tries to focus that window.
// Otherwise spawns a new terminal.
func OpenBranchInTerminal(branchName string) error {
	if !BranchSessionExists(branchName) {
		return fmt.Errorf("no tmux session for %s", branchName)
	}

	session := BranchSessionName(branchName)

	// If already attached, try to focus the existing window
	if BranchHasAttachedClients(branchName) {
		if focusTerminalByTitle(session) {
			return nil
		}
		// Couldn't focus, will spawn new terminal anyway
	}

	return spawnTerminalForSession(session)
}

// focusTerminalByTitle tries to focus a terminal window by title.
func focusTerminalByTitle(title string) bool {
	// Try xdotool (Linux)
	if _, err := exec.LookPath("xdotool"); err == nil {
		cmd := exec.Command("xdotool", "search", "--name", title)
		out, err := cmd.Output()
		if err == nil && len(strings.TrimSpace(string(out))) > 0 {
			windowID := strings.Split(strings.TrimSpace(string(out)), "\n")[0]
			exec.Command("xdotool", "windowactivate", windowID).Run()
			return true
		}
	}

	// Try wmctrl (Linux)
	if _, err := exec.LookPath("wmctrl"); err == nil {
		if exec.Command("wmctrl", "-a", title).Run() == nil {
			return true
		}
	}

	// macOS: try osascript (less reliable for specific windows)
	if _, err := exec.LookPath("osascript"); err == nil {
		script := `tell application "System Events" to set frontmost of (first process whose name contains "Terminal" or name contains "iTerm") to true`
		exec.Command("osascript", "-e", script).Run()
	}

	return false
}

// spawnTerminalForSession spawns a new terminal window attached to a tmux session.
func spawnTerminalForSession(session string) error {
	terminal := detectTerminal()
	attachCmd := fmt.Sprintf("tmux attach -t %s", session)

	var cmd *exec.Cmd
	switch terminal {
	case "gnome-terminal":
		cmd = exec.Command("gnome-terminal", "--title", session, "--", "bash", "-c", attachCmd)
	case "kitty":
		cmd = exec.Command("kitty", "--title", session, "bash", "-c", attachCmd)
	case "alacritty":
		cmd = exec.Command("alacritty", "--title", session, "-e", "bash", "-c", attachCmd)
	case "hyper":
		cmd = exec.Command("hyper", attachCmd)
	case "iterm2":
		script := fmt.Sprintf(`tell application "iTerm2"
			create window with default profile command "%s"
		end tell`, attachCmd)
		cmd = exec.Command("osascript", "-e", script)
	case "terminal":
		script := fmt.Sprintf(`tell application "Terminal"
			do script "%s"
			activate
		end tell`, attachCmd)
		cmd = exec.Command("osascript", "-e", script)
	case "wezterm":
		cmd = exec.Command("wezterm", "start", "--", "bash", "-c", attachCmd)
	case "xterm":
		cmd = exec.Command("xterm", "-title", session, "-e", attachCmd)
	default:
		return fmt.Errorf("unknown terminal: %s. Set DARK_MULTI_TERMINAL", terminal)
	}

	return cmd.Start()
}

// detectTerminal returns the configured or auto-detected terminal.
func detectTerminal() string {
	if config.Terminal != "auto" {
		return config.Terminal
	}

	terminals := []string{
		"kitty", "alacritty", "gnome-terminal", "hyper", "wezterm", "xterm",
	}

	// macOS defaults
	if _, err := exec.LookPath("osascript"); err == nil {
		if _, err := os.Stat("/Applications/iTerm.app"); err == nil {
			return "iterm2"
		}
		return "terminal"
	}

	// Linux: check what's installed
	for _, t := range terminals {
		if _, err := exec.LookPath(t); err == nil {
			return t
		}
	}

	return "xterm"
}

// Legacy compatibility - these wrap the new per-branch functions

// SessionExists returns true if any dark session exists (legacy).
func SessionExists() bool {
	// Check for any dark-* session
	out, err := exec.Command("tmux", "list-sessions", "-F", "#{session_name}").Output()
	if err != nil {
		return false
	}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.HasPrefix(line, "dark-") {
			return true
		}
	}
	return false
}

// WindowExists is deprecated - use BranchSessionExists instead.
func WindowExists(name string) bool {
	return BranchSessionExists(name)
}

// CreateWindow is deprecated - use CreateBranchSession instead.
func CreateWindow(name string, containerID string, branchPath string) error {
	return CreateBranchSession(name, containerID, branchPath)
}

// KillWindow is deprecated - use KillBranchSession instead.
func KillWindow(name string) error {
	return KillBranchSession(name)
}
